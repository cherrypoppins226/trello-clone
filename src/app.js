import React from "react";
import { withStyles } from "material-ui/styles";
import Grid from "material-ui/Grid";
import Button from "material-ui/Button";
import Paper from "material-ui/Paper";
import List, { ListItem, ListItemText } from "material-ui/List";
import PropTypes from "prop-types";
import red from "material-ui/colors/red";
import grey from "material-ui/colors/grey";
import Create from "@material-ui/icons/Create";

const CardsListCardStyles = {
  container: {
    margin: 8,
    marginTop: 0,
    marginBottom: 8,
    background: red[100],
    display: "flex",
    flexDirection: "row"
  },
  card: {
    padding: 4,
    paddingLeft: 8,
    "&:hover $edit": {
      visibility: "visible"
    }
  },
  text: {
    paddingRight: 0
  },
  edit: {
    visibility: "hidden",
    alignSelf: "flex-start",
    borderRadius: 4,
    padding: 2,
    paddingBottom: 0,
    "&:hover": {
      background: red[200]
    }
  },
  editIcon: {
    padding: 2,
    width: "0.8em",
    height: "0.8em",
    color: grey[800]
  }
};

const CardsListCardComponent = ({ classes, text = "Title..." }) => {
  return (
    <Paper elevation={1} component="li" className={classes.container}>
      <ListItem button className={classes.card}>
        <ListItemText className={classes.text}>{text}</ListItemText>
        <a className={classes.edit}>
          <Create className={classes.editIcon} />
        </a>
      </ListItem>
    </Paper>
  );
};

CardsListCardComponent.propTypes = {
  children: PropTypes.node,
  classes: PropTypes.object.isRequired,
  text: PropTypes.string
};

export const CardsListCard = withStyles(CardsListCardStyles)(
  CardsListCardComponent
);

const CardsListStyles = {
  container: {
    display: "flex",
    flexDirection: "column",
    maxHeight: "100%",
    background: red[50]
  },
  list: {
    padding: 1,
    overflowY: "scroll"
  },
  name: {
    padding: 0,
    margin: 8,
    marginTop: 16
  },
  addCard: {
    justifyContent: "left",
    textTransform: "none"
  }
};

class CardsListComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      cards: props.cards.map((text, idx) => (
        <CardsListCard key={idx} text={text} />
      ))
    };
  }

  addCard(e) {
    e.preventDefault();
    this.setState((prevState, props) => {
      const card = <CardsListCard key={this.state.cards.length + 1} />;
      return { cards: prevState.cards.concat(card) };
    });
  }

  render() {
    const { classes, name } = this.props;
    return (
      <Paper elevation={1} className={classes.container}>
        <h4 className={classes.name}>{name}</h4>
        <List data-testid="cards-list" className={classes.list}>
          {this.state.cards}
        </List>
        <Button className={classes.addCard} onClick={this.addCard.bind(this)}>
          Add a card...
        </Button>
      </Paper>
    );
  }
}

CardsListComponent.propTypes = {
  children: PropTypes.node,
  classes: PropTypes.object.isRequired,
  name: PropTypes.string.isRequired,
  cards: PropTypes.array.isRequired
};

export const CardsList = withStyles(CardsListStyles)(CardsListComponent);

const BoardStyles = {
  grid: {
    padding: 8,
    margin: 0,
    height: "100%",
    width: "100%"
  },
  cardListContainer: {
    width: 300
  }
};

const BoardComponent = ({ classes, lists }) => {
  return (
    <Grid container spacing={16} wrap="nowrap" className={classes.grid}>
      {Object.entries(lists).map(([k, v]) => {
        return (
          <Grid item key={k} className={classes.cardListContainer}>
            <CardsList name={k} cards={v} />
          </Grid>
        );
      })}
    </Grid>
  );
};

BoardComponent.propTypes = {
  children: PropTypes.node,
  classes: PropTypes.object.isRequired,
  lists: PropTypes.object.isRequired
};

export const Board = withStyles(BoardStyles)(BoardComponent);

const App = ({ lists }) => <Board lists={lists} />;

export default App;
