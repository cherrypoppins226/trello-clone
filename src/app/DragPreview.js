/*
 * Rendered as any draggable component in the app is being dragged
 *
 * There's only one instance of this component in the entire application, so it
 * needs to know how to render previews for all draggable components.
 */

import React from "react";
import { DragLayer } from "react-dnd";

import { CardDragPreview } from "./cardsList/DraggableCard";

const renderComponent = (componentType, itemBeingDragged) => {
  switch (componentType) {
    case "Card":
      return <CardDragPreview {...itemBeingDragged} />;
    default:
      return null;
  }
};

const componentStyles = props => {
  if (!props.currentOffset) return { display: "none" };
  const { x, y } = props.currentOffset;
  const transform = `translate(${x}px, ${y}px)`;
  return {
    position: "fixed",
    top: 0,
    left: 0,
    pointerEvents: "none",
    transform,
    WebkitTransform: transform
  };
};

export default DragLayer(monitor => ({
  itemBeingDragged: monitor.getItem(),
  componentType: monitor.getItemType(),
  currentOffset: monitor.getSourceClientOffset(),
  beingDragged: monitor.isDragging()
}))(
  class CustomDragLayer extends React.Component {
    render() {
      if (!this.props.beingDragged) return null;
      return (
        <div style={componentStyles(this.props)}>
          {renderComponent(
            this.props.componentType,
            this.props.itemBeingDragged
          )}
        </div>
      );
    }
  }
);
