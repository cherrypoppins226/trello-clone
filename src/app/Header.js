import React from "react";
import { withStyles } from "@material-ui/core/styles";

import logo from "../logo.png";

const styles = {
  root: {
    display: "flex",
    padding: 4,
    height: 40,
    justifyContent: "space-around",
    background: "#026AA7",
    "& img": {
      opacity: 0.5,
      width: 80,
      height: 30
    }
  }
};

const Header = ({ classes }) => (
  <div className={classes.root}>
    <img src={logo} alt="logo" />
  </div>
);

const Component = withStyles(styles)(Header);

export default Component;
