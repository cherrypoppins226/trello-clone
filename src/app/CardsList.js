import React from "react";
import { withStyles } from "@material-ui/core/styles";
import Button from "@material-ui/core/Button";
import Paper from "@material-ui/core/Paper";
import PropTypes from "prop-types";
import { fileAbsolute } from "paths.macro";
import { graphql } from "react-apollo";
import { compose, setDisplayName, setPropTypes } from "recompose";

import AppState from "../App.state";
import Header from "./cardsList/Header";
import Cards from "./cardsList/Cards";
import { queries } from "../cosmos/apollo/schema";
import { button } from "./styles";
import { makeFixtures, handleGraphQLResponse, moduleName } from "../utils";

const styles = {
  root: {
    display: "flex",
    flexDirection: "column",
    background: "rgb(226, 228, 230)",
    "& > button": {
      ...button,
      justifyContent: "left",
      fontWeight: 400
    }
  },
  header: {
    margin: 5
  }
};

const CardsList = ({ classes, id, addCard, data: { list, variables } }) => {
  return (
    <Paper
      data-listid={id}
      component="section"
      elevation={1}
      className={classes.root}
    >
      <Header
        className={classes.header}
        listId={list.id}
        listTitle={list.title}
      />
      <div style={{ overflowY: "scroll" }}>
        <Cards cards={list.cards} />
      </div>
      <Button
        onClick={() => {
          const listQuery = { query: queries.list, variables };
          addCard({
            update: (proxy, { data: { addCard } }) => {
              const data = proxy.readQuery(listQuery);
              data.list.cards.push(addCard);
              proxy.writeQuery({ ...listQuery, data });
            }
          });
        }}
      >
        Add a card...
      </Button>
    </Paper>
  );
};

const Component = compose(
  setDisplayName(moduleName(fileAbsolute)),
  setPropTypes({
    id: PropTypes.number.isRequired
  }),
  graphql(queries.list, {
    options: props => ({ variables: { id: props.id } })
  }),
  handleGraphQLResponse(),
  graphql(queries.addCard, {
    name: "addCard",
    options: props => ({
      variables: { listId: props.id, title: "Title..." },
      optimisticResponse: {
        __typename: "Mutation",
        addCard: {
          __typename: "Card",
          id: -1,
          title: "Title..."
        }
      }
    })
  }),
  withStyles(styles)
)(CardsList);

export const fixtures = makeFixtures(Component, {
  default: {
    props: {
      id: 1
    },
    stores: {
      appState: new AppState()
    }
  }
});

export default Component;
