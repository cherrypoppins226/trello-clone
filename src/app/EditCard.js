import React from "react";
import { findDOMNode } from "react-dom";
import { withStyles } from "@material-ui/core/styles";
import { fileAbsolute } from "paths.macro";
import { compose } from "recompose";
import { inject } from "mobx-react";

import { makeFixtures, labelId, moduleName } from "../utils";
import Header from "./editCard/Header";
import Sidebar from "./editCard/Sidebar";
import Content from "./editCard/Content";

const modulePath = moduleName(fileAbsolute);

export const labels = {
  description: {
    id: labelId(modulePath, "description"),
    text: "Edit card: Change title, descriptions, labels, attachments, and more"
  }
};

const styles = {
  root: {
    width: 750,
    backgroundColor: "rgb(237, 239, 240)",
    borderRadius: 2,
    outline: "none"
  },
  flex: {
    display: "flex"
  },
  content: {
    width: "100%",
    padding: "0px 10px"
  },
  sidebar: {
    flexBasis: 220
  }
};

// TODO: Shouldn't be a concern here
const centerInPage = node => {
  node.style.left = `${(window.innerWidth - node.offsetWidth) / 2}px`;
};

class EditCard extends React.Component {
  componentDidMount() {
    this.node = findDOMNode(this);
    centerInPage(this.node);
  }

  render() {
    const { classes, appState, ...rest } = this.props;
    return (
      <div
        aria-describedby={labels.description.id}
        className={classes.root}
        tabIndex={-1}
        {...rest}
      >
        <Header
          className={classes.header}
          cardId={appState.cardBeingEdited.id}
          cardTitle={appState.cardBeingEdited.title}
        />
        <div className={classes.flex}>
          <Content className={classes.content} />
          <Sidebar className={classes.sidebar} />
        </div>
      </div>
    );
  }
}

const Component = compose(inject("appState"), withStyles(styles))(EditCard);

export const fixtures = makeFixtures(modulePath, Component, {
  default: {
    props: {},
    stores: {
      appState: {
        cardBeingEdited: { id: 1, title: "Ut sunt qui amet." },
        finishCardEdit: () => {}
      }
    }
  }
});

export default Component;
