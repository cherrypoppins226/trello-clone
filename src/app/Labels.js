import React from "react";
import { withStyles } from "@material-ui/core/styles";

import { flatten } from "../utils";

const labels = flatten(
  [
    require("./EditCard").labels,
    require("./QuickEditCard").labels,
    require("./cardsList/ActionsMenu").labels,
    require("./cardsList/Card").labels,
    require("./cardsList/Header").labels,
    require("./cardsList/NewCard").labels
  ].map(labels => Object.values(labels))
);

// https://webaim.org/techniques/css/invisiblecontent/
const styles = {
  root: {
    position: "absolute",
    left: -10000,
    top: "auto",
    width: 1,
    height: 1,
    overflow: "hidden"
  }
};

const Labels = ({ classes }) => (
  <div className={classes.root}>
    {labels.map(({ id, text }) => (
      <div key={id} id={id}>
        {text}
      </div>
    ))}
  </div>
);

const Component = withStyles(styles)(Labels);

export default Component;
