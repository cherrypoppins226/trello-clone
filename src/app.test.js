import React from "react";
import { shallow } from "enzyme";
import faker from "faker";
import { Board, CardsList, CardsListCard } from "./app";

faker.seed(1);

const lists = {
  [faker.lorem.sentence()]: [faker.lorem.sentence(), faker.lorem.sentence()],
  [faker.lorem.sentence()]: [faker.lorem.sentence(), faker.lorem.sentence()]
};

describe("renders correctly", () => {
  it("board", () => {
    const tree = shallow(<Board lists={lists} />).dive();
    expect(tree).toMatchSnapshot();
  });

  it("cards-list", () => {
    const [name, cards] = Object.entries(lists)[0];
    const tree = shallow(<CardsList name={name} cards={cards} />).dive();
    expect(tree).toMatchSnapshot();
  });

  it("cards-list card", () => {
    const tree = shallow(<CardsListCard text="Card list card text" />).dive();
    expect(tree).toMatchSnapshot();
  });
});
