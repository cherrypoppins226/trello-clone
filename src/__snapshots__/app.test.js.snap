// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`board initial render 1`] = `
<WithStyles(GridWrapper)
  className="BoardComponent-grid-1"
  container={true}
  spacing={16}
  wrap="nowrap"
>
  <WithStyles(GridWrapper)
    className="BoardComponent-cardListContainer-2"
    item={true}
    key="Repellat quisquam recusandae alias consequuntur corporis."
  >
    <WithStyles(CardsListComponent)
      cards={
        Array [
          "Ratione ut sunt qui amet iure ut libero qui recusandae.",
          "Nulla quam ipsam nobis cupiditate sed dignissimos.",
        ]
      }
      name="Repellat quisquam recusandae alias consequuntur corporis."
    />
  </WithStyles(GridWrapper)>
  <WithStyles(GridWrapper)
    className="BoardComponent-cardListContainer-2"
    item={true}
    key="Incidunt accusantium sed libero repudiandae esse blanditiis natus et eos."
  >
    <WithStyles(CardsListComponent)
      cards={
        Array [
          "Velit omnis et porro ut et ipsam explicabo eligendi occaecati.",
          "Et saepe eum dicta eum eaque enim ipsum inventore debitis.",
        ]
      }
      name="Incidunt accusantium sed libero repudiandae esse blanditiis natus et eos."
    />
  </WithStyles(GridWrapper)>
</WithStyles(GridWrapper)>
`;

exports[`cards list card initial render 1`] = `
<WithStyles(Paper)
  className="CardsListCardComponent-container-7"
  component="li"
  elevation={1}
>
  <WithStyles(ListItem)
    button={true}
    className="CardsListCardComponent-card-8"
  >
    <WithStyles(ListItemText)
      className="CardsListCardComponent-text-9"
    >
      Card list card text
    </WithStyles(ListItemText)>
    <a
      className="CardsListCardComponent-edit-10"
    >
      <pure(Create)
        className="CardsListCardComponent-editIcon-11"
      />
    </a>
  </WithStyles(ListItem)>
</WithStyles(Paper)>
`;

exports[`cards list initial render 1`] = `
<WithStyles(Paper)
  className="CardsListComponent-container-3"
  elevation={1}
>
  <h4
    className="CardsListComponent-name-5"
  >
    Repellat quisquam recusandae alias consequuntur corporis.
  </h4>
  <WithStyles(List)
    className="CardsListComponent-list-4"
    data-testid="cards-list"
  >
    <WithStyles(CardsListCardComponent)
      key="0"
      text="Ratione ut sunt qui amet iure ut libero qui recusandae."
    />
    <WithStyles(CardsListCardComponent)
      key="1"
      text="Nulla quam ipsam nobis cupiditate sed dignissimos."
    />
  </WithStyles(List)>
  <WithStyles(Button)
    className="CardsListComponent-addCard-6"
    onClick={[Function]}
  >
    Add a card...
  </WithStyles(Button)>
</WithStyles(Paper)>
`;
